### PR TITLE
Redirection wasb

### PR DESCRIPTION
**Pull request by anmolanmol1234**
This PR adds the functionality to redirect to wasb for renames and delete. The redirection to rename is false by default and the redirection to delete is true by default. Since wasb with oauth support was not available earlier, this PR also has the code changes to enable wasb with oauth authentication method.

### CODE REVIEW

**Comment by saxenapranav** [comment]
please give more about what pr is, what are we changing. Thanks.

**Comment by anmolanmol1234** [comment]
> please give more about what pr is, what are we changing. Thanks.

Updated PR description.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java [hunk 1]
```diff
@@ -347,6 +347,38 @@ public AbfsConfiguration(final Configuration rawConfig, String accountName)
...
+
+  @VisibleForTesting
+  public void setRedirectRename(boolean redirectRename) {
+    this.redirectRename = redirectRename;
+  }
+
+  @VisibleForTesting
+  public void setRedirectDelete(boolean redirectDelete) {
+    this.redirectDelete = redirectDelete;
+  }
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: not used. lets remove it; atleast not have it public.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azure/AzureBlobStorageTestAccount.java [hunk 2]
```diff
@@ -21,6 +21,7 @@
 import com.microsoft.azure.storage.*;
 import com.microsoft.azure.storage.blob.*;
```
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azure/AzureBlobStorageTestAccount.java:22]: test cases to assert rename and delete pass with redirection enabled needed.

Comment (snvijaya) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azure/AzureBlobStorageTestAccount.java:22]: These tests are needed only for nonHNS + 0Auth , combination, test can check and skip run if these conditions are not qualified.

**Review by snvijaya — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java [hunk 3]
```diff
@@ -540,6 +607,32 @@ public boolean rename(final Path src, final Path dst) throws IOException {
...
 
+    TracingContext tracingContext = new TracingContext(clientCorrelationId,
+            fileSystemId, FSOperationType.RENAME, true, tracingHeaderFormat,
+            listener);
+
+    if (shouldRedirect(FSOperationType.RENAME, tracingContext)) {
+      LOG.debug("Rename redirected for the given src {} and the given destination {}", src, dst);
+      Path wasbSrc = src;
+      Path wasbDest = dst;
+      if (src.toString().contains(FileSystemUriSchemes.ABFS_SCHEME)
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: should we have 
```
if(FileSystemUriSchemes.ABFS_SCHEME.equalsIgnoreCase(src.toUri().getScheme())) {}
```

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/FileSystemConfigurations.java [hunk 4]
```diff
@@ -119,6 +119,9 @@ public final class FileSystemConfigurations {
   public static final int STREAM_ID_LEN = 12;
   public static final boolean DEFAULT_ENABLE_ABFS_LIST_ITERATOR = true;
 
+  public static final boolean DEFAULT_FS_AZURE_REDIRECT_RENAME = false;
+  public static final boolean DEFAULT_FS_AZURE_REDIRECT_DELETE = true;
+
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/FileSystemConfigurations.java]: lets add a comment on why a one config is true and why other is false.

**Review by saxenapranav — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azure/AzureBlobStorageTestAccount.java (hunk 2 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azure/AzureBlobStorageTestAccount.java:22]: Added test cases but not specified to run only for this specific combination, as want to validate that redirection works with all auth types

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java (hunk 1 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: Done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java (hunk 3 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java]: Done

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/FileSystemConfigurations.java (hunk 4 above)
Comment (anmolanmol1234) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/constants/FileSystemConfigurations.java]: Done

**Review by snvijaya — APPROVED** [review]
LGTM

**Review by saxenapranav — APPROVED** [review]
LGTM!

**Pull request merged**